### ISSUE TITLE
understanding nested CV results

### ISSUE THREAD

**Issue by AleBitetto** [mlr-org/mlr#2627]
Hi,

I'm using nested CV for parameter tuning as described in the tutorial. I have the following questions:

1 - If I correctly understood, in each of the outer training set we find the best parameter set, so in the end the final best parameter set is chosen among tuning results of different outer train sets. Shouldn't be better to test the same parameter set, according to the aggregation measure, evaluated on every outer train/test set? In case, is there a way to evaluate this using the easy parallelized framework `mlr` provides?

2 - if I use the following measure `meas = list(rmse, setAggregation(rmse, test.sd), setAggregation(rmse, train.mean), setAggregation(rmse, train.sd))`, are the mean and std I get in the `getNestedTuneResultsOptPathDf` related on a single outer train/test pair (reported in the `iter` column), aggregating the inner CV results? Then, how `r$measures.test` and `r$measures.train` are evaluated? Are they evaluated in each outer train/test pair using the model with the best param set chosen in the corresponding outer trainig set?

Thanks a lot in advance!

**Comment by pat-s** [comment]
Hi,

theoretical questions on understanding basic ML concepts are usually better asked at [Cross-Validated](https://stats.stackexchange.com/) or looked up in literature.

1.No. You want to find the "best" parameter set for the current "set" you're having. What "best" means here is defined by the optimization measure during tuning.

> Shouldn't be better to test the same parameter set, according to the aggregation measure, evaluated on every outer train/test set?

This is exactly what you do not want to do because this is not a fair comparison. You never benchmark settings on the same set you are applying them to later on (which would be here your outer test set); instead you always use a subset of the train data.

> are the mean and std I get in the getNestedTuneResultsOptPathDf related on a single outer train/test pair (reported in the iter column), aggregating the inner CV results?

Yes, the aggregation function sets the way how the single fold error measures are aggregated after resampling.

> Then, how r$measures.test and r$measures.train are evaluated? Are they evaluated in each outer train/test pair using the model with the best param set chosen in the corresponding outer trainig set?

Yes, for each fold you get a "train" and "test" error measure by predicting your trained model (with the tuned parameters) either on the training or test set. Note that the train set estimates are always better and do not reflect the real performance of your model. Always go with the test set estimates.

**Comment by AleBitetto** [comment]
Thank a lot for the answer.

Probably I wasn't clear enough on point 1. I'm saying that if we choose the best parameter set among the best parameter sets _PS_i_ in each outer fold _OF_i_ we may discard (or overestimate) some _PS_i_ just because we were (un)lucky in training them in _OF_i_ training set. I was thinking on evaluating the _PS_i_ estimated in _OF_i_ also in every other _OF_j_, _i!=j_, (training and testing with _PS_i_ on _OF_j_) in order to see how stable the parameters are averaging over all _OF_k_ test set. So I'm saying that we should get an average (or whatever aggregation measure) for each entry of `r$measures.test`.

Hope this is much clearer. Thanks in advance

**Comment by berndbischl** [comment]
hi this really is not even a question about the usability of mlr anymore.
what mlr has implemented is absolut standard procedure, documented in books an papers. 
we cannot turn the tracker into a discussion forum for general ML matters, sory.
